### ISSUE TITLE
Empty storable vectors in GHCi error with divide by zero.

### ISSUE THREAD

**Issue by mihaimaruseac** [haskell/vector#167]
In GHCi:

```haskell
=> import Data.Vector.Storable
=> empty
*** Exception: divide by zero
```

Enabling `-XNoMonomorphismRestriction`, giving it an explicit type or having it compiled by `ghc` from a file makes the error go away.

This doesn't happen for either boxed (`Data.Vector`) or unboxed (`Data.Vector.Unboxed`) vectors.

I consider it a `vector` bug since it's an inconsistency between the unboxed and the storable case. But, if it is not, feel free to close this/raise it to the proper bug tracker.

**Comment by RyanGlScott** [comment]
Well, this issue didn't exist in GHC 7.10.3 or earlier by sheer coincidence. The reason it changed in 8.0 is that a new `Storable ()` instance was added:

```haskell
instance Storable () where
  sizeOf _ = 0
  alignment _ = 1
  peek _ = return ()
  poke _ _ = return ()
```

And luck just has it that if you type `empty` at the REPL with no type annotations, its type will default to `Vector ()`. It seems that having `sizeOf _ = 0` causes the divide by zero error. I'm not familiar enough with the codebase to know _where_ the faulty division is happening, though.

**Comment by RyanGlScott** [comment]
A quick search reveals [this](https://github.com/haskell/vector/blob/49396e1e54084ce9fd3982b99030ca316ad05856/Data/Vector/Storable/Mutable.hs#L131-L132) as a likely culprit:

```haskell
instance Storable a => G.MVector MVector a where
  ...
  basicUnsafeNew n
    | n < 0 = error $ "Storable.basicUnsafeNew: negative length: " ++ show n
    | n > mx = error $ "Storable.basicUnsafeNew: length too large: " ++ show n
    | otherwise = unsafePrimToPrim $ do
        fp <- mallocVector n
        return $ MVector n fp
    where
      size = sizeOf (undefined :: a)
      mx = maxBound `quot` size :: Int
```

**Comment by mihaimaruseac** [comment]
Makes sense and indeed that seems to be the culprit.

Now, for solution, the simplest way would be to disallow making a storable vector of `()`. At least this will give educative error messages. I made #168 for this, as a starting point.

**Comment by cartazio** [comment]
I think there's a preexisting ticket about handling zero width storable
values.

Making them erroneous probably isn't the right approach ...

**Comment by mihaimaruseac** [comment]
Yeah, it's issue #105 (I only searched for it now). I see that no progress has been made on it :(

I changed my PR so now it divides by 1 if the size of the elements is 0.

**Comment by mihaimaruseac** [comment]
Compiled the following with profiling

```haskell
import qualified Data.Vector.Storable as VS

main = print (VS.empty :: VS.Vector ())
```

and indeed got the expected error

```haskell
*** Exception (reporting due to +RTS -xc): (THUNK_STATIC), stack trace:
  GHC.Real.CAF
  --> evaluated by: Data.Vector.Storable.Mutable.basicUnsafeNew,
  called from Main.main,
  called from Main.CAF:main3
  --> evaluated by: Main.main,
  called from Main.CAF:main2
  --> evaluated by: Main.main
mm: divide by zero
```

The error goes away with fix I proposed in #168

**Comment by dolio** [comment]
Fix merged.
